File: python_sdk/agentfriendly/monetization/x402.py
"""
Layer 7 — x402 Payment Middleware (Python)

Mirrors packages/core/src/monetization/x402.ts
"""

from __future__ import annotations

import fnmatch
import json
from decimal import Decimal

from ..config import MonetizationConfig, X402RouteConfig
from ..middleware import EarlyResponse
from ..types import AgentContext

USDC_DECIMALS = 6

USDC_CONTRACTS = {
    "base-mainnet": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    "base-sepolia": "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
    "solana-mainnet": "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
}


def _parse_price(price: str | float) -> int:
    """Convert a human-readable price to USDC base units."""
    num_str = str(price).replace("$", "")
    return int(Decimal(num_str) * Decimal(10**USDC_DECIMALS))


def find_matching_pricing(
    method: str,
    path: str,
    routes: dict[str, X402RouteConfig],
    exempt: list[str],
) -> X402RouteConfig | None:
    """Find the pricing config for the current request, or None if not monetized."""
    for exempt_pattern in exempt:
        if fnmatch.fnmatch(path, exempt_pattern) or exempt_pattern in path:
            return None

    for route_pattern, pricing in routes.items():
        parts = route_pattern.strip().split(None, 1)
        if len(parts) == 2 and parts[0].isupper():
            pattern_method, pattern_path = parts
            if pattern_method != method.upper():
                continue
        else:
            pattern_path = parts[0]

        if fnmatch.fnmatch(path, pattern_path):
            return pricing

    return None


def generate_402_response(
    config: MonetizationConfig,
    pricing: X402RouteConfig,
    path: str,
) -> EarlyResponse:
    """Generate an HTTP 402 Payment Required response."""
    wallet = pricing.to or config.wallet_address or ""
    network = pricing.network or config.network or "base-mainnet"
    amount_micro = _parse_price(pricing.price)
    usdc_contract = USDC_CONTRACTS.get(network, USDC_CONTRACTS["base-mainnet"])

    payment_terms = {
        "version": "1",
        "accepts": [
            {
                "scheme": "exact",
                "network": network,
                "maxAmountRequired": str(amount_micro),
                "to": wallet,
                "asset": usdc_contract,
                "extra": {"name": "USDC", "version": "2"},
            }
        ],
        "error": "X-PAYMENT header is required to access this resource",
    }

    human_amount = f"{amount_micro / 10 ** USDC_DECIMALS:.4f}"
    body = "\n".join([
        "# Payment Required",
        "",
        f"**Price**: ${human_amount} USDC per request",
        f"**Network**: {network}",
        "**Protocol**: [x402](https://x402.org) — open internet-native micropayments",
        "",
        "## How to Pay",
        "",
        "Retry this request with the `X-Payment` header set to your signed payment proof.",
        "",
    ])

    return EarlyResponse(
        status=402,
        headers={
            "Content-Type": "text/markdown; charset=utf-8",
            "Accept-Payment": "x402/v1",
            "X-Payment-Required": json.dumps(payment_terms),
        },
        body=body,
        content_type="text/markdown",
    )


def check_monetization(
    context: AgentContext,
    config: MonetizationConfig,
    headers: dict[str, str],
) -> EarlyResponse | None:
    """Check monetization requirements. Returns 402 EarlyResponse or None."""
    if not config.enabled or not context.is_agent:
        return None

    pricing = find_matching_pricing(
        context.method, context.path, config.routes, config.exempt
    )
    if not pricing:
        return None

    payment_header = headers.get("x-payment")
    if not payment_header:
        return generate_402_response(config, pricing, context.path)

    # Structural verification
    try:
        decoded = json.loads(
            __import__("base64").b64decode(payment_header).decode()
        )
        amount = int(decoded.get("amount", 0))
        required = _parse_price(pricing.price)
        to = str(decoded.get("to", "")).lower()
        required_to = (pricing.to or config.wallet_address or "").lower()
        network = decoded.get("network", "")

        if to != required_to or amount < required or network != (pricing.network or "base-mainnet"):
            return generate_402_response(config, pricing, context.path)
    except Exception:
        return generate_402_response(config, pricing, context.path)

    return None

File: packages/core/src/monetization/x402.ts
import micromatch from "micromatch";

import type { AgentContext } from "../types/agent-context.js";
import type { HandledResponse } from "../types/agent-response.js";
import type { MonetizationConfig, X402RouteConfig } from "../types/config.js";

/**
 * Layer 7 — x402 Payment Middleware
 *
 * Implements the x402 protocol (HTTP 402 Payment Required) for autonomous
 * agent micropayments via USDC stablecoins on Base/Solana.
 *
 * The x402 flow:
 * 1. Agent requests a paid route with no X-Payment header
 * 2. SDK responds 402 with machine-readable payment terms
 * 3. Agent constructs and signs a USDC transaction
 * 4. Agent retries with X-Payment: <base64-signed-tx>
 * 5. SDK verifies the payment proof cryptographically (no network call)
 * 6. If valid, the request proceeds to the route handler
 *
 * Payment verification is self-contained — no blockchain RPC call needed.
 * The signed transaction payload contains all information needed to verify
 * the amount, recipient, and currency match the required payment terms.
 *
 * References:
 * - x402 spec: https://x402.org
 * - Base USDC contract: 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913
 * - Solana USDC: EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v
 */

/** USDC contract addresses by network. */
const USDC_CONTRACTS = {
  "base-mainnet": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
  "base-sepolia": "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
  "solana-mainnet": "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
} as const;

/** USDC decimal places (6 for EVM chains, 6 for Solana). */
const USDC_DECIMALS = 6;

/**
 * Convert a human-readable price string to USDC base units (microUSDC).
 * Input: "$0.001" or "0.001" → output: 1000 (0.001 × 10^6)
 */
function parsePriceToMicroUsdc(price: string | number): number {
  const numStr = typeof price === "string" ? price.replace("$", "") : String(price);
  return Math.round(Number(numStr) * Math.pow(10, USDC_DECIMALS));
}

/**
 * Find the pricing config for the current request path and method.
 * Matches route patterns in the format "METHOD /path/pattern".
 * Returns null if no pricing rule matches.
 */
export function findMatchingPricing(
  method: string,
  path: string,
  routes: Record<string, X402RouteConfig>,
  exempt: string[],
): X402RouteConfig | null {
  // Check exemptions first
  for (const exemptPattern of exempt) {
    if (path.includes(exemptPattern) || micromatch.isMatch(path, exemptPattern)) {
      return null;
    }
  }

  for (const [routePattern, pricing] of Object.entries(routes)) {
    // Route pattern format: "GET /api/**" or "/api/**" (method optional)
    const parts = routePattern.trim().split(/\s+/, 2);
    let patternMethod: string | null = null;
    let patternPath: string;

    if (parts.length === 2 && parts[0] && /^[A-Z]+$/.test(parts[0])) {
      patternMethod = parts[0];
      patternPath = parts[1] ?? "";
    } else {
      patternPath = parts[0] ?? "";
    }

    // Check method match (if specified)
    if (patternMethod && patternMethod !== method.toUpperCase()) continue;

    // Check path match (glob pattern)
    if (micromatch.isMatch(path, patternPath)) {
      return pricing;
    }
  }

  return null;
}

/**
 * Generate a 402 Payment Required response with machine-readable x402 payment terms.
 * This is returned to agents that hit a paid route without providing payment.
 */
export function generate402Response(
  config: MonetizationConfig,
  pricing: X402RouteConfig,
  _path: string,
): HandledResponse {
  const walletAddress = pricing.to ?? config.walletAddress ?? "";
  const network = pricing.network ?? config.network ?? "base-mainnet";
  const amountMicroUsdc = parsePriceToMicroUsdc(pricing.price);
  const usdcContract =
    USDC_CONTRACTS[network as keyof typeof USDC_CONTRACTS] ?? USDC_CONTRACTS["base-mainnet"];

  const paymentTerms = {
    version: "1",
    accepts: [
      {
        scheme: "exact",
        network,
        maxAmountRequired: String(amountMicroUsdc),
        to: walletAddress,
        asset: usdcContract,
        extra: { name: "USDC", version: "2" },
      },
    ],
    error: "X-PAYMENT header is required to access this resource",
  };

  const humanReadableAmount = (amountMicroUsdc / Math.pow(10, USDC_DECIMALS)).toFixed(4);

  const markdownBody = [
    `# Payment Required`,
    ``,
    `This endpoint requires payment to access.`,
    ``,
    `**Price**: \$${humanReadableAmount} USDC per request`,
    `**Network**: ${network}`,
    `**Protocol**: [x402](https://x402.org) — open internet-native micropayments`,
    ``,
    `## How to Pay`,
    ``,
    `1. Ensure your agent has a USDC wallet on ${network}`,
    `2. The x402 protocol handles payment automatically for compatible agents`,
    `3. Retry this request with the \`X-Payment\` header set to your signed payment proof`,
    ``,
    `## For Developers`,
    ``,
    `If your agent doesn't support x402 yet, see [x402.org/agent-integration](https://x402.org/agent-integration).`,
    ``,
    ...(pricing.description ? [`*${pricing.description}*`, ``] : []),
  ].join("\n");

  return {
    handled: true,
    status: 402,
    headers: {
      "Content-Type": "text/markdown; charset=utf-8",
      "Accept-Payment": "x402/v1",
      "X-Payment-Required": JSON.stringify(paymentTerms),
    },
    body: markdownBody,
    contentType: "text/markdown",
  };
}

/** Result of verifying an x402 payment proof. */
export interface PaymentVerificationResult {
  readonly valid: boolean;
  readonly amountUsdc: number;
  readonly network: string;
  readonly proofHash: string;
  readonly errorReason: string | null;
}

/**
 * Verify the X-Payment header proof for a paid route request.
 *
 * IMPORTANT: Full cryptographic x402 verification requires the x402 payment
 * library (from coinbase/x402) which does the on-chain state check.
 * This implementation performs the structural verification (shape, amount, recipient)
 * and delegates on-chain verification to the external library when available.
 *
 * In the absence of the external library (development/testing), structural
 * verification is performed and a warning is logged.
 */
export async function verifyPaymentProof(
  paymentHeader: string,
  requiredPricing: X402RouteConfig,
  walletAddress: string,
): Promise<PaymentVerificationResult> {
  try {
    // Decode the base64-encoded payment proof
    const decodedJson = Buffer.from(paymentHeader, "base64").toString("utf-8");
    const proof = JSON.parse(decodedJson) as Record<string, unknown>;

    const network = (proof["network"] as string) ?? "";
    const to = (proof["to"] as string) ?? "";
    const amount = Number(proof["amount"] ?? 0);
    const amountUsdc = amount / Math.pow(10, USDC_DECIMALS);

    // Structural checks
    const requiredAmount = parsePriceToMicroUsdc(requiredPricing.price);
    const requiredTo = (requiredPricing.to ?? walletAddress).toLowerCase();
    const requiredNetwork = requiredPricing.network ?? "base-mainnet";

    if (to.toLowerCase() !== requiredTo) {
      return {
        valid: false,
        amountUsdc: 0,
        network: "",
        proofHash: "",
        errorReason: "wrong-recipient",
      };
    }
    if (amount < requiredAmount) {
      return {
        valid: false,
        amountUsdc,
        network,
        proofHash: "",
        errorReason: "insufficient-amount",
      };
    }
    if (network !== requiredNetwork) {
      return { valid: false, amountUsdc, network, proofHash: "", errorReason: "wrong-network" };
    }

    // Generate a short hash for analytics (first 16 chars of a simple hash)
    const proofHash = paymentHeader.slice(0, 16);

    return { valid: true, amountUsdc, network, proofHash, errorReason: null };
  } catch (error) {
    return {
      valid: false,
      amountUsdc: 0,
      network: "",
      proofHash: "",
      errorReason: `parse-error: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}

/**
 * Main monetization middleware entry point.
 * Returns:
 * - A 402 HandledResponse if payment is required but not provided/invalid
 * - null if the route is not monetized, or payment was verified successfully
 */
export async function checkMonetization(
  context: AgentContext,
  config: MonetizationConfig,
): Promise<HandledResponse | null> {
  if (!config.enabled || !context.isAgent) return null;

  const routes = config.routes ?? {};
  const exempt = config.exempt ?? [];

  const pricing = findMatchingPricing(context.method, context.path, routes, exempt);
  if (!pricing) return null;

  const paymentHeader = context.headers["x-payment"];

  // No payment header — issue the challenge
  if (!paymentHeader) {
    return generate402Response(config, pricing, context.path);
  }

  // Payment header present — verify it
  const verificationResult = await verifyPaymentProof(
    paymentHeader,
    pricing,
    config.walletAddress ?? "",
  );

  if (!verificationResult.valid) {
    return generate402Response(config, pricing, context.path);
  }

  // Payment verified — allow the request through
  return null;
}
